[rpc] make handling of FORWARD_AUTOGRAD_REQ in request_callback_impl
nonblocking.

rohan-varma opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#32476 [rpc] make handling of FORWARD_AUTOGRAD_REQ in request_callback_impl
nonblocking.**
nonblocking.**
nonblocking.**

This makes the handling of FORWARD_AUTOGRAD_REQ in request_callback
nonblocking. Processing this message requires unwrapping the message with
autograd information, processing the original message, and sending back the
message with autograd information wrapped. This makes the processing the
original message nonblocking by grabbing a future to it and marking the parent
future as completed when this one completes.

Differential Revision: [D19509501](https://our.internmc.facebook.com/intern/diff/D19509501/)

rohan-varma committed:
[rpc] make handling of FORWARD_AUTOGRAD_REQ in request_callback_impl
nonblocking.

This makes the handling of FORWARD_AUTOGRAD_REQ in request_callback
nonblocking. Processing this message requires unwrapping the message with
autograd information, processing the original message, and sending back the
message with autograd information wrapped. This makes the processing the
original message nonblocking by grabbing a future to it and marking the parent
future as completed when this one completes.

Differential Revision: [D19509501](https://our.internmc.facebook.com/intern/diff/D19509501/)

[ghstack-poisoned]

rohan-varma committed:
Update on "[rpc] make handling of FORWARD_AUTOGRAD_REQ in request_callback_impl
nonblocking."

nonblocking.**

nonblocking.

This makes the handling of FORWARD_AUTOGRAD_REQ in request_callback
nonblocking. Processing this message requires unwrapping the message with
autograd information, processing the original message, and sending back the
message with autograd information wrapped. This makes the processing the
original message nonblocking by grabbing a future to it and marking the parent
future as completed when this one completes.

Differential Revision: [D19509501](https://our.internmc.facebook.com/intern/diff/D19509501/)

[ghstack-poisoned]

rohan-varma committed:
Update on "[rpc] make handling of FORWARD_AUTOGRAD_REQ in request_callback_impl
nonblocking."

nonblocking.**

This makes the handling of FORWARD_AUTOGRAD_REQ in request_callback
nonblocking. Processing this message requires unwrapping the message with
autograd information, processing the original message, and sending back the
message with autograd information wrapped. This makes the processing the
original message nonblocking by grabbing a future to it and marking the parent
future as completed when this one completes.

Differential Revision: [D19509501](https://our.internmc.facebook.com/intern/diff/D19509501/)

[ghstack-poisoned]

jjlilley commented on torch/csrc/distributed/rpc/request_callback_impl.cpp:
Instead of forcing a copy here, you can probably do something like  wrappedRpcResponseFuture->moveValue()

Also, I think you need to propagate any error

rohan-varma committed:
Update on "[rpc] make handling of FORWARD_AUTOGRAD_REQ in request_callback_impl
nonblocking."

nonblocking.**

This makes the handling of FORWARD_AUTOGRAD_REQ in request_callback
nonblocking. Processing this message requires unwrapping the message with
autograd information, processing the original message, and sending back the
message with autograd information wrapped. This makes the processing the
original message nonblocking by grabbing a future to it and marking the parent
future as completed when this one completes.

Differential Revision: [D19509501](https://our.internmc.facebook.com/intern/diff/D19509501/)

[ghstack-poisoned]

rohan-varma committed:
Update on "[rpc] make handling of FORWARD_AUTOGRAD_REQ in request_callback_impl
nonblocking."

nonblocking.**
nonblocking.**

This makes the handling of FORWARD_AUTOGRAD_REQ in request_callback
nonblocking. Processing this message requires unwrapping the message with
autograd information, processing the original message, and sending back the
message with autograd information wrapped. This makes the processing the
original message nonblocking by grabbing a future to it and marking the parent
future as completed when this one completes.

Differential Revision: [D19509501](https://our.internmc.facebook.com/intern/diff/D19509501/)

[ghstack-poisoned]

jjlilley commented on torch/csrc/distributed/rpc/request_callback_impl.cpp:
I think you probably want to setId(messageId) on the message before marking completed with it.

jjlilley commented on torch/csrc/distributed/rpc/request_callback_impl.cpp:
pls remove the var name 'm', to make the unused var linter issue go away.

jjlilley reviewed (commented):
Looks nice, some minor stuff I missed the first pass

rohan-varma commented on torch/csrc/distributed/rpc/request_callback_impl.cpp:
@jjlilley I'll add this, but could you clarify the reasoning why we need to do this? I couldn't find where the messageId was previously set when returning a wrapped message with `getMessageWithAutograd`, so I'm not sure what changes here would require us to set it.

jjlilley commented on torch/csrc/distributed/rpc/request_callback_impl.cpp:
I'm suggestion doing so, simply because we were calling setId() previously, via the wrap() lambda call.

(If you go back to December, before we returned a future from this method, there was a function that ensured that any Message had the id set).

I honestly don't know if it's strictly necessary to do so, but it seems safer to keep this aspect of behavior as-is, unless there's a reason to change this.

rohan-varma committed:
Update on "[rpc] make handling of FORWARD_AUTOGRAD_REQ in request_callback_impl
nonblocking."

nonblocking.**

This makes the handling of FORWARD_AUTOGRAD_REQ in request_callback
nonblocking. Processing this message requires unwrapping the message with
autograd information, processing the original message, and sending back the
message with autograd information wrapped. This makes the processing the
original message nonblocking by grabbing a future to it and marking the parent
future as completed when this one completes.

Differential Revision: [D19509501](https://our.internmc.facebook.com/intern/diff/D19509501/)

[ghstack-poisoned]

rohan-varma commented on torch/csrc/distributed/rpc/request_callback_impl.cpp:
Added in the latest change.

kostmo commented:
## :pill: CircleCI build failures summary and remediations
As of commit 5f6c5e27:


* **1/1** failures introduced in this PR


## Detailed failure analysis

One may explore the probable reasons each build failed interactively [on the Dr. CI website](https://dr.pytorch.org/commit-details.html?sha1=5f6c5e2742d2af3fdb35c38a181f36f88a0c7243).

### :detective: 1 new failure recognized by patterns
The following build failures do not appear to be due to upstream breakage:
#### [![See CircleCI build](https://avatars0.githubusercontent.com/ml/7?s=12)](https://circleci.com/gh/pytorch/pytorch/4326943) pytorch_macos_10_13_py3_test (1/1)
**Step:** "Test" ([full log](https://dr.pytorch.org/api/view-log-full?build_id=90010909) | [pattern match details](https://dr.pytorch.org/build-details.html?build_id=90010909))
<details>
<summary>
<code>Jan 23 12:39:17 RuntimeError: test_autograd failed!</code>
</summary>

```
Jan 23 12:39:17 Ran 1046 tests in 281.609s 
Jan 23 12:39:17  
Jan 23 12:39:17 FAILED (errors=1, skipped=13, expected failures=1) 
Jan 23 12:39:17  
Jan 23 12:39:17 Generating XML reports... 
Jan 23 12:39:17 Traceback (most recent call last): 
Jan 23 12:39:17   File "test/run_test.py", line 456, in <module> 
Jan 23 12:39:17     main() 
Jan 23 12:39:17   File "test/run_test.py", line 449, in main 
Jan 23 12:39:17     raise RuntimeError(message) 
Jan 23 12:39:17 RuntimeError: test_autograd failed! 
Jan 23 12:39:17 + cleanup 
Jan 23 12:39:17 + retcode=1 
Jan 23 12:39:17 + set +x 
```

</details>



---

<details><summary>This comment was automatically generated by <a href="https://github.com/kostmo/circleci-failure-tracker/tree/master/docs/from-pull-request-comment">Dr. CI</a> (expand for details).</summary>Follow <a href="https://dr.pytorch.org/admin/comments-opt-out.html">this link to opt-out</a> of these comments for your Pull Requests.

Please report bugs/suggestions on the <a href="https://github.com/kostmo/circleci-failure-tracker/issues">GitHub issue tracker</a>.
</details>

jjlilley commented on torch/csrc/distributed/rpc/request_callback_impl.cpp:
minor: can you get rid of the std::move() here?

(see the lint on the Phabricator side)

I think the issue is that without the std::move(), the compiler can use a vanilla return-value optimization (NRVO), basically the result just gets constructed in the caller stack by the callee, without any copy.

When you add the std::move(), I suspect you force an extra, unnecessary move constructor here.

That said, I think you might also want to put a std::move() in the markCompleted() statement, a few lines down.

rohan-varma committed:
Update on "[rpc] make handling of FORWARD_AUTOGRAD_REQ in request_callback_impl
nonblocking."

nonblocking.**
nonblocking.**

This makes the handling of FORWARD_AUTOGRAD_REQ in request_callback
nonblocking. Processing this message requires unwrapping the message with
autograd information, processing the original message, and sending back the
message with autograd information wrapped. This makes the processing the
original message nonblocking by grabbing a future to it and marking the parent
future as completed when this one completes.

Differential Revision: [D19509501](https://our.internmc.facebook.com/intern/diff/D19509501/)

[ghstack-poisoned]

jjlilley commented:
LGTM!
One oddity - I don't see the unittest results on the PR page (but a message about "confirm squash and merge")

jjlilley commented:
Oh, weird, now that I actually said that the tests weren't showing up, they are, nevermind :)

rohan-varma commented:
@jjlilley Yeah I've seen that happen sometimes, not sure what's causing it. If you're using FaceHub, Pieter posted an internal issue about this happening a few months ago.

Could you stamp the PR if it looks good?

jjlilley reviewed (approved):
LGTM!

rohan-varma committed:
Update on "[rpc] make handling of FORWARD_AUTOGRAD_REQ in request_callback_impl
nonblocking."

nonblocking.**

This makes the handling of FORWARD_AUTOGRAD_REQ in request_callback
nonblocking. Processing this message requires unwrapping the message with
autograd information, processing the original message, and sending back the
message with autograd information wrapped. This makes the processing the
original message nonblocking by grabbing a future to it and marking the parent
future as completed when this one completes.

Differential Revision: [D19509501](https://our.internmc.facebook.com/intern/diff/D19509501/)

[ghstack-poisoned]

rohan-varma committed:
Update on "[rpc] make handling of FORWARD_AUTOGRAD_REQ in request_callback_impl
nonblocking."

nonblocking.**
nonblocking.**

This makes the handling of FORWARD_AUTOGRAD_REQ in request_callback
nonblocking. Processing this message requires unwrapping the message with
autograd information, processing the original message, and sending back the
message with autograd information wrapped. This makes the processing the
original message nonblocking by grabbing a future to it and marking the parent
future as completed when this one completes.

Differential Revision: [D19509501](https://our.internmc.facebook.com/intern/diff/D19509501/)

[ghstack-poisoned]

rohan-varma commented:
Test failures are due to flakiness - reported https://github.com/pytorch/pytorch/issues/32644, https://github.com/pytorch/pytorch/issues/32643
